File: lib/model/entities/weight.dart
/*
   Copyright 2020 Thomas Bonk <thomas@meandmymac.de>

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

import 'package:floor/floor.dart';
import 'package:flutter/foundation.dart';

import 'package:repti/model/entities/index.dart';
import 'base_entity.dart';

@Entity(
  tableName: 'weight',
  foreignKeys: [
    ForeignKey(
      childColumns: ['individual_id'],
      parentColumns: ['id'],
      entity: Individual,
      onDelete: ForeignKeyAction.cascade,
    )
  ],
)
class Weight extends BaseEntity {
  @ColumnInfo(name: 'individual_id')
  final String individualId;

  @ColumnInfo(name: 'weight')
  double weight;

  Weight({
    @required Individual individual,
    @required double weight,
    String id,
    DateTime updateTime,
    DateTime createTime,
  })  : this.individualId = individual.id,
        this.weight = weight,
        super(id, updateTime, createdAt: createTime);
}

File: Repti/Source/Persistence/Model/Weight.swift
//
//  Weight.swift
//  Repti
//
//  Created by Thomas Bonk on 05.01.21.
//  Copyright 2021 Thomas Bonk <thomas@meandmymac.de>
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

import Foundation
import CoreData

protocol WeightProtocol {
  var date: Date { get set }
  var weight: Float { get set }
}

extension Weight: WeightProtocol {

  // MARK: - Public Properties

  public var dao: WeightDAO {
    return WeightDAO(date: self.date, weight: self.weight)
  }

  // MARK: - Class Methods

  class func create(in managedObjectContext: NSManagedObjectContext) -> Weight {
    return
      NSEntityDescription
      .insertNewObject(
        forEntityName: Weight.entityName,
        into: managedObjectContext) as! Weight
  }
}

struct WeightDAO: WeightProtocol {

  // MARK: - Public Properties

  var date: Date
  var weight: Float


  // MARK: - Initialization

  init(date: Date, weight: Float) {
    self.date = date
    self.weight = weight
  }

  init(weighing: Weight) {
    self.date = weighing.date
    self.weight = weighing.weight
  }
}
